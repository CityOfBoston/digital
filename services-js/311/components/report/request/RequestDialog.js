// @flow

import React from 'react';
import Head from 'next/head';
import { css } from 'glamor';

import type { Service, SubmittedRequest, Request } from '../../../data/types';
import type { LoopbackGraphql } from '../../../data/graphql/loopback-graphql';
import { submitRequest } from '../../../data/store/request';

import FormDialog from '../../common/FormDialog';
import SectionHeader from '../../common/SectionHeader';

import { QuestionsPaneContainer, ContactPaneContainer, LocationPopUpContainer } from './containers';
import SubmitPane from './SubmitPane';

type ExternalProps = {
  service: ?Service,
  stage: 'questions' | 'location' | 'contact',
  loopbackGraphql: LoopbackGraphql,
  routeToServiceForm: (code: string, stage: string) => void,
  setLocationMapActive: (active: boolean) => void,
  locationMapSearch: ?(query: string) => Promise<boolean>,
};

export type ValueProps = {
  request: Request,
}

export type ActionProps = {
  resetRequestForService: (service: Service) => void,
}

type Props = ExternalProps & ValueProps & ActionProps;

const COMMON_DIALOG_STYLE = {
  transition: 'margin 400ms',
};

const CENTERED_DIALOG_STYLE = css(COMMON_DIALOG_STYLE, {
});

const CORNER_DIALOG_STYLE = css(COMMON_DIALOG_STYLE, {
  margin: '-60px 60% 0 20px',
});

/*
  <LocationPopUp address={address} addressSearch={this.whenAddressSearch} next={this.whenNextClicked} />
*/

export default class RequestDialog extends React.Component {
  props: Props;
  state: {
    submitting: boolean,
    submitError: ?Object,
    submittedRequest: ?SubmittedRequest,
  }

  constructor(props: Props) {
    super(props);

    this.state = {
      submitting: false,
      submitError: null,
      submittedRequest: null,
    };
  }

  componentWillMount() {
    const { service, resetRequestForService, stage, setLocationMapActive } = this.props;
    if (service) {
      resetRequestForService(service);
    }

    setLocationMapActive(stage === 'location');
  }

  componentWillReceiveProps(newProps: Props) {
    const { setLocationMapActive, stage } = this.props;

    if (stage !== newProps.stage) {
      setLocationMapActive(newProps.stage === 'location');
    }
  }

  nextAfterQuestions = () => {
    const { service, routeToServiceForm } = this.props;
    if (service) {
      routeToServiceForm(service.code, 'location');
    }
  }

  nextAfterLocation = () => {
    const { service, routeToServiceForm } = this.props;
    if (service) {
      routeToServiceForm(service.code, 'contact');
    }
  }

  nextAfterContact = () => {
    this.submitRequest();
  }

  submitRequest = async () => {
    // Pull the store out of the context rather than via connect because we
    // don't need it for rendering, just for doing this submission.
    this.setState({
      submitting: true,
      submitError: null,
      submittedRequest: null,
    });

    try {
      const { request, loopbackGraphql } = this.props;
      const { createRequest } = await submitRequest(request, loopbackGraphql);

      this.setState({
        submitting: false,
        submittedRequest: createRequest,
      });
    } catch (e) {
      this.setState({
        submitError: e,
      });
    }
  }

  render() {
    const { stage } = this.props;
    const dialogContainerStyle = stage === 'location' ? CORNER_DIALOG_STYLE : CENTERED_DIALOG_STYLE;

    return (
      <div className={dialogContainerStyle}>
        <Head>
          <title>BOS:311 — {this.renderTitle()}</title>
        </Head>

        <FormDialog>
          { this.renderContent() }
        </FormDialog>
      </div>
    );
  }

  renderTitle() {
    const { stage, service } = this.props;
    const { submitting, submitError, submittedRequest } = this.state;

    if (!service) {
      return 'Not Found';
    }

    if (submitError) {
      return 'Submission error';
    } else if (submitting) {
      return 'Submitting…';
    } else if (submittedRequest) {
      return `Success: Case ${submittedRequest.id}`;
    }

    switch (stage) {
      case 'questions': return service.name;
      case 'location': return 'Choose location';
      case 'contact': return 'Contact information';
      default: return '';
    }
  }

  renderContent() {
    const { service, stage, locationMapSearch } = this.props;
    const { submitting, submitError, submittedRequest } = this.state;

    if (submitError) {
      return <SubmitPane state="error" error={submitError} />;
    } else if (submitting) {
      return <SubmitPane state="submitting" />;
    } else if (submittedRequest) {
      return <SubmitPane state="success" submittedRequest={submittedRequest} />;
    }

    if (!service) {
      return <SectionHeader>Service not found</SectionHeader>;
    }

    switch (stage) {
      case 'questions':
        return <QuestionsPaneContainer nextFunc={this.nextAfterQuestions} />;

      case 'location':
        return <LocationPopUpContainer nextFunc={this.nextAfterLocation} addressSearch={locationMapSearch} />;

      case 'contact':
        return <ContactPaneContainer service={service} nextFunc={this.nextAfterContact} />;

      default: return null;
    }
  }
}
