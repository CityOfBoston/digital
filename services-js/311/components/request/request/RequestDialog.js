// @flow

import React from 'react';
import Head from 'next/head';
import Router from 'next/router';
import { action, observable, autorun } from 'mobx';
// import ScopedError from 'react-scoped-error-component';
import { observer } from 'mobx-react';
import { fromPromise } from 'mobx-utils';
import { css } from 'glamor';
import type { Context } from 'next';
import type { IPromiseBasedObservable } from 'mobx-utils';

import type { RequestAdditions } from '../../../server/next-handlers';

import type { SubmittedRequest, Service } from '../../../data/types';
import type { AppStore } from '../../../data/store';
import RequestForm from '../../../data/store/RequestForm';
import makeLoopbackGraphql from '../../../data/dao/loopback-graphql';
import type { LoopbackGraphql } from '../../../data/dao/loopback-graphql';
import submitRequest from '../../../data/dao/submit-request';
import loadService from '../../../data/dao/load-service';

import LoadingBuildings from '../../common/LoadingBuildings';
import FormDialog from '../../common/FormDialog';
import SectionHeader from '../../common/SectionHeader';

import { MEDIA_LARGE, CENTERED_DIALOG_STYLE } from '../../style-constants';

import QuestionsPane from './QuestionsPane';
import LocationPopUp from './LocationPopUp';
import ContactPane from './ContactPane';
import SubmitPane from './SubmitPane';
import CaseView from '../../case/CaseView';

export type InitialProps = {|
  stage: 'questions' | 'location' | 'contact' | 'submit',
  service: ?Service,
  description: string,
|};

export type Props = {|
  store: AppStore,
  loopbackGraphql: LoopbackGraphql,
  routeToServiceForm: (code: string, stage: string) => Promise<void>,
  setLocationMapActive: (active: boolean) => void,
  ...InitialProps,
|};

const COMMON_DIALOG_STYLE = {
  transition: 'margin 400ms',
  position: 'relative',
  zIndex: 2,
};

export const CORNER_DIALOG_STYLE = css(COMMON_DIALOG_STYLE, {
  margin: 0,
  [MEDIA_LARGE]: {
    // comedy bottom margin to push this up into the corner, otherwise it would
    // get vertically centered.
    margin: '-0px 70% 90vh 40px',
    minWidth: 450,
  },
});

@observer
export default class RequestDialog extends React.Component {
  props: Props;

  requestForm: RequestForm;
  @observable
  requestSubmission: ?IPromiseBasedObservable<SubmittedRequest> = null;

  caseLookupOnSubmitDisposer: Function;

  // Called by RequestLayout
  static async getInitialProps({
    query,
    req,
    res,
  }: Context<RequestAdditions>): Promise<InitialProps> {
    const { code, description } = query;
    let stage = query.stage;

    const loopbackGraphql = makeLoopbackGraphql(req);

    const service = await loadService(loopbackGraphql, code);

    if (!service && res) {
      res.statusCode = 404;
    }

    stage = stage || 'questions';

    switch (stage) {
      case 'questions':
      case 'location':
      case 'contact':
      case 'submit':
        return {
          service,
          stage,
          description: description || '',
        };
      default:
        throw new Error(`Unknown stage: ${stage}`);
    }
  }

  @action
  componentWillMount() {
    const { stage, description, setLocationMapActive, service } = this.props;

    this.requestForm = new RequestForm(service);
    this.requestForm.description = description;

    setLocationMapActive(stage === 'location');
  }

  componentDidMount() {
    LoadingBuildings.preload();

    this.caseLookupOnSubmitDisposer = autorun('route on case submit', () => {
      if (
        this.requestSubmission &&
        this.requestSubmission.state === 'fulfilled'
      ) {
        const { id } = this.requestSubmission.value;
        // We keep the first URL the same to avoid a page load, but change the
        // "as" to match the URL for the case permalink.
        Router.replace('/request', `/reports/${id}`, {
          shallow: true,
        });
      }
    });
  }

  componentWillUnmount() {
    this.caseLookupOnSubmitDisposer();
  }

  @action
  componentWillReceiveProps(newProps: Props) {
    const { setLocationMapActive, stage } = this.props;

    if (stage !== newProps.stage) {
      setLocationMapActive(newProps.stage === 'location');
    }
  }

  makeNextAfterQuestions(): { fn: () => mixed, isSubmit: boolean } {
    const { requestForm } = this;

    if (requestForm.showLocationPicker) {
      return {
        fn: this.routeToLocation,
        isSubmit: false,
      };
    } else if (requestForm.showContactInfoForm) {
      return {
        fn: this.routeToContact,
        isSubmit: false,
      };
    } else {
      return {
        fn: this.submitRequest,
        isSubmit: true,
      };
    }
  }

  makeNextAfterLocation(): { fn: () => mixed, isSubmit: boolean } {
    const { requestForm } = this;

    if (requestForm.showContactInfoForm) {
      return {
        fn: this.routeToContact,
        isSubmit: false,
      };
    } else {
      return {
        fn: this.submitRequest,
        isSubmit: true,
      };
    }
  }

  routeToLocation = () => {
    const { service, routeToServiceForm } = this.props;
    if (service) {
      routeToServiceForm(service.code, 'location');
    }
  };

  routeToContact = () => {
    const { service, routeToServiceForm } = this.props;
    if (service) {
      routeToServiceForm(service.code, 'contact');
    }
  };

  @action.bound
  submitRequest(): Promise<mixed> {
    const { requestForm } = this;
    const { service, loopbackGraphql, routeToServiceForm } = this.props;
    const {
      description,
      firstName,
      lastName,
      email,
      phone,
      location,
      address,
      addressId,
      questions,
      mediaUrl,
      sendLocation,
      sendContactInfo,
    } = requestForm;

    if (!service) {
      throw new Error('service is null in submitRequest');
    }

    const promise = submitRequest(loopbackGraphql, {
      service,
      description,
      // We only send contact info with an affirmative boolean that's set by
      // the ContactPane form's submit button. That way contact info that was
      // loaded by localStorage won't be pushed to the server if the user didn't
      // see the contact info form or chose "submit without sending contact info"
      firstName: sendContactInfo ? firstName : null,
      lastName: sendContactInfo ? lastName : null,
      email: sendContactInfo ? email : null,
      phone: sendContactInfo ? phone : null,
      location: sendLocation ? location : null,
      address: sendLocation ? address : null,
      addressId: sendLocation ? addressId : null,
      questions,
      mediaUrl,
    }).then(
      v => {
        // upload was successful, so clear out previous request state
        this.requestForm = new RequestForm(service);
        return v;
      },
      err => {
        /* eslint-disable no-underscore-dangle */
        if (window._opbeat) {
          window._opbeat('captureException', err);
          err._sentToOpbeat = true;
        }
        /* eslint-enable no-underscore-dangle */

        // This exception is ultimately "caught" by the requestSubmission fromPromise
        // mobx handler.
        throw err;
      }
    );

    this.requestSubmission = fromPromise(promise);

    routeToServiceForm(service.code, 'submit');

    return promise;
  }

  render() {
    const { stage } = this.props;
    const dialogContainerStyle =
      stage === 'location'
        ? CORNER_DIALOG_STYLE
        : css(COMMON_DIALOG_STYLE, CENTERED_DIALOG_STYLE);

    return (
      <div className={dialogContainerStyle}>
        <Head>
          <title>
            BOS:311 — {this.renderTitle()}
          </title>
        </Head>

        <FormDialog
          narrow={stage === 'contact'}
          noPadding={stage === 'location'}
        >
          {this.renderContent()}
        </FormDialog>
      </div>
    );
  }

  renderTitle() {
    const { requestSubmission } = this;
    const { stage, service } = this.props;

    if (!service) {
      return 'Not Found';
    }

    switch (stage) {
      case 'questions':
        return service.name;
      case 'location':
        return 'Choose location';
      case 'contact':
        return 'Contact information';
      case 'submit':
        if (!requestSubmission) {
          return '';
        }

        switch (requestSubmission.state) {
          case 'pending':
            return 'Submitting…';
          case 'fulfilled':
            return `Success: Case ${requestSubmission.value.id}`;
          case 'rejected':
            return 'Submission error';
          default:
            return '';
        }

      default:
        return '';
    }
  }

  renderContent() {
    const { requestSubmission, requestForm } = this;
    const { store, stage, service } = this.props;

    if (!service) {
      return <SectionHeader>Service not found</SectionHeader>;
    }

    // TODO(finh): Bring back service errors
    // if (currentServiceError) {
    //   return (
    //     <div>
    //       <SectionHeader>Error loading service</SectionHeader>
    //       <ScopedError error={currentServiceError} />
    //     </div>
    //   );
    // }

    switch (stage) {
      case 'questions': {
        const { fn, isSubmit } = this.makeNextAfterQuestions();
        return (
          <QuestionsPane
            store={store}
            requestForm={requestForm}
            serviceName={service.name}
            serviceDescription={service.description}
            nextFunc={fn}
            nextIsSubmit={isSubmit}
          />
        );
      }

      case 'location': {
        const { fn, isSubmit } = this.makeNextAfterLocation();
        return (
          <LocationPopUp
            store={store}
            requestForm={requestForm}
            nextFunc={fn}
            nextIsSubmit={isSubmit}
          />
        );
      }

      case 'contact':
        return (
          <ContactPane
            requestForm={requestForm}
            serviceName={service.name}
            nextFunc={this.submitRequest}
          />
        );

      case 'submit':
        if (!requestSubmission) {
          return '';
        }

        switch (requestSubmission.state) {
          case 'pending':
            return <SubmitPane state="submitting" ui={store.ui} />;
          case 'fulfilled':
            return (
              <CaseView
                store={store}
                request={requestSubmission.value}
                submitted
              />
            );
          case 'rejected':
            return <SubmitPane state="error" error={requestSubmission.value} />;
          default:
            return '';
        }

      default:
        return null;
    }
  }
}
