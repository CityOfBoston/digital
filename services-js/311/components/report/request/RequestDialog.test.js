// @flow

import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import RequestDialog from './RequestDialog';

import type { Service, SubmittedRequest } from '../../../data/types';
import { AppStore } from '../../../data/store';

jest.mock('../../../data/dao/submit-request');
const submitRequest: JestMockFn = (require('../../../data/dao/submit-request'): any).default;

const MOCK_SERVICE: Service = {
  name: 'Cosmic Incursion',
  code: 'CSMCINC',
  contactRequired: true,
  locationRequired: true,
  attributes: [{
    required: false,
    type: 'TEXT',
    code: 'ST-CMTS',
    description: 'Please provide any other relevant information:',
    values: null,
    conditionalValues: null,
    dependencies: null,
  }, {
    required: false,
    type: 'STRING',
    code: 'INFO-CSIRMV1',
    description: '**All cosmic incursion cases should be followed up with a phone call to Alpha Flight.**',
    values: null,
    conditionalValues: null,
    dependencies: null,
  }, {
    required: true,
    type: 'SINGLEVALUELIST',
    code: 'SR-CSIRMV1',
    description: 'How many dimensions have been breached?',
    values: [{ key: 'One', name: 'One' }, { key: 'Two', name: 'Two' }, { key: 'Three', name: 'Three' }, { key: 'More than Three', name: 'More than Three' }],
    conditionalValues: [],
    dependencies: null,
  }],
};

const MOCK_ACTIONS = {
  loopbackGraphql: jest.fn(),
  routeToServiceForm: jest.fn(),
  resetRequestForService: jest.fn(),
  setLocationMapActive: jest.fn(),
  locationMapSearch: jest.fn(),
};

describe('rendering', () => {
  let store;

  beforeEach(() => {
    store = new AppStore();
    store.currentService = MOCK_SERVICE;
  });

  test('questions', () => {
    const component = renderer.create(
      <RequestDialog store={store} stage="questions" {...MOCK_ACTIONS} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('location', () => {
    const component = renderer.create(
      <RequestDialog store={store} stage="location" {...MOCK_ACTIONS} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('contact', () => {
    const component = renderer.create(
      <RequestDialog store={store} stage="contact" {...MOCK_ACTIONS} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('methods', () => {
  let wrapper;
  let store;
  let requestDialog;
  const routeToServiceForm = jest.fn();

  beforeEach(() => {
    store = new AppStore();
    store.currentService = MOCK_SERVICE;

    wrapper = shallow(
      <RequestDialog
        store={store}
        stage="submit"
        locationMapSearch={jest.fn()}
        loopbackGraphql={jest.fn()}
        routeToServiceForm={routeToServiceForm}
        setLocationMapActive={jest.fn()}
      />);

    requestDialog = wrapper.instance();
  });

  describe('navigation', () => {
    test('nextAfterQuestions', () => {
      requestDialog.nextAfterQuestions();
      expect(routeToServiceForm).toHaveBeenCalledWith(MOCK_SERVICE.code, 'location');
    });

    test('nextAfterLocation', () => {
      requestDialog.nextAfterLocation();
      expect(routeToServiceForm).toHaveBeenCalledWith(MOCK_SERVICE.code, 'contact');
    });

    test('nextAfterContact', () => {
      requestDialog.submitRequest = jest.fn();
      requestDialog.nextAfterContact();
      expect(requestDialog.submitRequest).toHaveBeenCalledWith();
    });
  });

  describe('submitRequest', () => {
    let resolveGraphql;
    let rejectGraphql;

    beforeEach(() => {
      const promise = new Promise((resolve, reject) => {
        resolveGraphql = resolve;
        rejectGraphql = reject;
      });

      submitRequest.mockReturnValue(promise);
    });

    test('success', async () => {
      const submission = requestDialog.submitRequest();

      // rendering loading
      wrapper.update();
      expect(wrapper).toMatchSnapshot();

      const result: SubmittedRequest = {
        id: '17-00001615',
        requestedAtString: 'January 1, 2017 2:00 PM',
        description: 'A cosmic encursion seems likely.',
        address: 'City Hall Plaza, Boston, MA',
        mediaUrl: 'https://pbs.twimg.com/media/C22X9ODXgAABGKS.jpg',
      };

      resolveGraphql(result);

      await submission;

      // rendering success
      wrapper.update();
      expect(wrapper).toMatchSnapshot();

      expect(store.description).toEqual('');
      expect(store.mediaUrl).toEqual('');
    });

    test('graphql failure', async () => {
      const submission = requestDialog.submitRequest();

      const error: Object = new Error('Error submitting');
      error.errors = [
        { message: 'All required fields were missing' },
        { message: 'Also your location is in Cambridge' },
      ];

      rejectGraphql(error);

      try {
        await submission;
      } catch (e) {
        // expected
      }

      // should be rendering the error here
      wrapper.update();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
