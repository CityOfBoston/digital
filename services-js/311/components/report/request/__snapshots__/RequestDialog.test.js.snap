// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`methods submitRequest graphql failure 1`] = `
<div
  className={
    Object {
      "data-css-12u9but": "",
    }
  }
>
  <SideEffect(Head)>
    <title>
      BOS:311 — 
      Submission error
    </title>
  </SideEffect(Head)>
  <FormDialog
    narrow={false}
    popup={false}
  >
    <SubmitPane
      error={[Error: Error submitting]}
      state="error"
    />
  </FormDialog>
</div>
`;

exports[`methods submitRequest success 1`] = `
<div
  className={
    Object {
      "data-css-12u9but": "",
    }
  }
>
  <SideEffect(Head)>
    <title>
      BOS:311 — 
      Submitting…
    </title>
  </SideEffect(Head)>
  <FormDialog
    narrow={false}
    popup={false}
  >
    <SubmitPane
      state="submitting"
      ui={
        Ui {
          "handleResize": [Function],
          "handleScroll": [Function],
        }
      }
    />
  </FormDialog>
</div>
`;

exports[`methods submitRequest success 2`] = `
<div
  className={
    Object {
      "data-css-12u9but": "",
    }
  }
>
  <SideEffect(Head)>
    <title>
      BOS:311 — 
      Success: Case 17-000000001
    </title>
  </SideEffect(Head)>
  <FormDialog
    narrow={false}
    popup={false}
  >
    <CaseView
      request={
        Object {
          "address": "City Hall Plaza, Boston, MA 02131",
          "description": "I think that Thanos is here",
          "id": "17-000000001",
          "location": Object {
            "lat": 42.359927299999995,
            "lng": -71.0576853,
          },
          "mediaUrl": "https://pbs.twimg.com/media/C22X9ODXgAABGKS.jpg",
          "requestedAtString": "March 7, 2017, 12:59 PM",
          "service": Object {
            "name": "Cosmic Intervention",
          },
          "status": "closed",
          "statusNotes": "Found Thanos. Smashed him into the floor with all of us standing around.",
          "updatedAtString": "April 8, 2017, 12:59 PM",
        }
      }
      store={
        AppStore {
          "_liveAgentButtonId": "",
          "accessibility": Accessibility {
            "el": null,
            "messageListenerDisposer": null,
          },
          "allServices": AllServices {
            "groups": Array [
              Group {
                "description": "Pet Ownership, Problems with Animals, Rodents & Pests",
                "id": "Animal Control",
                "name": "Animals",
              },
              Group {
                "description": "Certificates & Registrations, Licensing & Permitting, Complaints, Business with the City",
                "id": "business",
                "name": "Business",
              },
              Group {
                "description": "City Jobs, Schools, Library Resources, Internships & Volunteering",
                "id": "education-employment",
                "name": "Education & Employment",
              },
              Group {
                "description": "Voting, Contact Elected Officials, Personal Records (e.g., Births, Deaths & Marriages)",
                "id": "elections-records",
                "name": "Elections & Records",
              },
              Group {
                "description": "Trash, Recycling, Graffiti, Dumping, Street Cleaning, Neighborhood Cleanup",
                "id": "garbage",
                "name": "Garbage, Recycling & Graffiti",
              },
              Group {
                "description": "Crime, Fire Prevention & Control, Healthy Living & Inspections",
                "id": "Health - Safety",
                "name": "Health, Safety & Social Services",
              },
              Group {
                "description": "Home Ownership, Renting, Inspections, Code Enforcement, Permitting, Complaints",
                "id": "housing",
                "name": "Housing, Property & Construction",
              },
              Group {
                "description": "Park Maintenance, Permits, Tree Planting, Donations",
                "id": "parks",
                "name": "Parks & Trees",
              },
              Group {
                "description": "Calendar, Event Permitting, Facility Request, Youth, Arts",
                "id": "entertainment",
                "name": "Recreation, Entertainment & Event Planning",
              },
              Group {
                "description": "Real Estate Tax, Boat and Motor Vehicle Excise, Personal Property",
                "id": "taxes",
                "name": "Taxes & Payments",
              },
              Group {
                "description": "Signs & Signals, Parking, Driving, Public Transportation, Taxi Issues, Road Maintenance",
                "id": "transportation",
                "name": "Transportation, Streets & Sidewalks",
              },
              Group {
                "description": "Water & Sewer, Electric, Gas, Heat",
                "id": "utilities",
                "name": "Utilities",
              },
            ],
            "otherGroup": Group {
              "description": "These services have not yet been categorized",
              "id": "",
              "name": "Other Services",
            },
          },
          "apiKeys": Object {},
          "browserLocation": BrowserLocation {
            "watchId": null,
          },
          "languages": Array [],
          "mapLocation": MapLocation {},
          "requestSearch": RequestSearch {
            "searchDisposer": null,
          },
          "ui": Ui {
            "handleResize": [Function],
            "handleScroll": [Function],
          },
        }
      }
      submitted={true}
    />
  </FormDialog>
</div>
`;

exports[`rendering contact 1`] = `
<div
  className={
    Object {
      "data-css-12u9but": "",
    }
  }
>
  <div
    className="p-a300 p-a800--xl br br-t400 br--y css-182uick"
    style={
      Object {
        "maxWidth": 800,
      }
    }
  >
    <form
      onSubmit={[Function]}
    >
      <div>
        <div
          className="sh sh--y"
        >
          <h1
            className="sh-title"
          >
            Cosmic Incursion
          </h1>
        </div>
        <p
          className="m-v300 t--info"
        >
          We’ll use your contact info to send you email about the status of your report and to follow up with you if necessary.
        </p>
        <p
          className="m-v300 t--subinfo"
        >
          <strong>
            Your contact info will not be made public.
          </strong>
           
        </p>
        <div
          className="txt"
        >
          <label
            className="txt-l"
            htmlFor="ContactPane-firstName"
          >
            First Name 
            <span
              className="t--req"
            >
              Required
            </span>
          </label>
          <input
            aria-required={true}
            className="txt-f"
            id="ContactPane-firstName"
            name="firstName"
            onChange={[Function]}
            placeholder="First Name"
            style={
              Object {
                "width": "100%",
              }
            }
            type="text"
            value=""
          />
        </div>
        <div
          className="txt"
        >
          <label
            className="txt-l"
            htmlFor="ContactPane-lastName"
          >
            Last Name 
            <span
              className="t--req"
            >
              Required
            </span>
          </label>
          <input
            aria-required={true}
            className="txt-f"
            id="ContactPane-lastName"
            name="lastName"
            onChange={[Function]}
            placeholder="Last Name"
            style={
              Object {
                "width": "100%",
              }
            }
            type="text"
            value=""
          />
        </div>
        <div
          className="txt"
        >
          <label
            className="txt-l"
            htmlFor="ContactPane-email"
          >
            Email 
            <span
              className="t--req"
            >
              Required
            </span>
          </label>
          <input
            aria-required={true}
            className="txt-f"
            id="ContactPane-email"
            name="email"
            onChange={[Function]}
            placeholder="Email"
            style={
              Object {
                "width": "100%",
              }
            }
            type="email"
            value=""
          />
        </div>
        <div
          className="txt"
        >
          <label
            className="txt-l"
            htmlFor="ContactPane-phone"
          >
            Phone
          </label>
          <input
            className="txt-f"
            id="ContactPane-phone"
            name="phone"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            onPaste={[Function]}
            placeholder="Phone"
            style={
              Object {
                "width": "100%",
              }
            }
            type="tel"
            value=""
          />
        </div>
        <div
          className="m-v500 g"
        >
          <div
            className="cb"
          >
            <input
              checked={false}
              className="cb-f"
              id="ContactPane-remember"
              name="remember"
              onChange={[Function]}
              type="checkbox"
              value="true"
            />
            <label
              className="cb-l"
              htmlFor="ContactPane-remember"
            >
              Remember contact info on this computer
            </label>
          </div>
        </div>
      </div>
      <div
        className="g m-v500 css-1k8tu3j"
      >
        <div
          className="g--8 t--info m-v200 css-hn0yet"
        >
          <span>
            Please fill out 
            <span
              className="t--req"
            >
              required
            </span>
             fields to continue
          </span>
        </div>
        <button
          className="btn g--4"
          disabled={true}
          type="submit"
        >
          Submit Report
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`rendering location 1`] = `
<div
  className={
    Object {
      "data-css-5ju6dk": "",
    }
  }
>
  <div
    className="p-a300 p-a500 br br-t400 br--y css-182uick"
    style={Object {}}
  >
    <div
      className={
        Object {
          "data-css-1szwbq4": "",
        }
      }
    >
      <form
        action="#"
        className="sf sf--sm sf--y "
        onSubmit={[Function]}
      >
        <div
          className="sf-i"
        >
          <input
            aria-label="Address search field"
            className="sf-i-f"
            onInput={[Function]}
            placeholder="Search for a street address or intersection…"
            type="text"
            value=""
          />
          <button
            className="sf-i-b"
            disabled={true}
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
      <div
        className="m-v400"
      />
      <div
        className="g css-1k8tu3j"
      >
        <div
          className="g--7 t--subinfo m-v200"
        >
          <a
            href="javascript:void(0)"
            onClick={[Function]}
          >
            Continue without location
          </a>
        </div>
        <button
          className="btn g--5"
          disabled={true}
          onClick={[Function]}
        >
          Next
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`rendering questions 1`] = `
<div
  className={
    Object {
      "data-css-12u9but": "",
    }
  }
>
  <div
    className="p-a300 p-a800--xl br br-t400 br--y css-182uick"
    style={Object {}}
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="sh sh--y"
      >
        <h1
          className="sh-title"
        >
          Cosmic Incursion
        </h1>
      </div>
      <div
        className="g g--top"
      >
        <div
          className="g--7 m-v500"
        >
          <div
            className="txt"
          >
            <label
              className="txt-l"
              htmlFor="QuestionsPane-description"
              style={
                Object {
                  "marginTop": 0,
                }
              }
            >
              Description
            </label>
            <textarea
              className="txt-f"
              id="QuestionsPane-description"
              name="description"
              onInput={[Function]}
              placeholder="Please describe your problem"
              rows="5"
              value=""
            />
          </div>
          <div>
            <div
              className="txt"
            >
              <label
                className="txt-l"
                htmlFor="ST-CMTS"
              >
                Please provide any other relevant information:
                 
              </label>
              <textarea
                aria-required={false}
                className="txt-f"
                id="ST-CMTS"
                name="ST-CMTS"
                onChange={[Function]}
                rows="5"
                value=""
              />
            </div>
          </div>
          <hr
            aria-hidden={true}
            className="hr hr--dash m-v500"
          />
          <div>
            <div
              className="txt"
            >
              <label
                className="txt-l"
                htmlFor="INFO-CSIRMV1"
              >
                **All cosmic incursion cases should be followed up with a phone call to Alpha Flight.**
                 
              </label>
              <input
                aria-required={false}
                className="txt-f"
                id="INFO-CSIRMV1"
                name="INFO-CSIRMV1"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
          </div>
          <hr
            aria-hidden={true}
            className="hr hr--dash m-v500"
          />
          <div>
            <div
              aria-labelledby="SR-CSIRMV1-label"
              role="radiogroup"
            >
              <div
                className="m-v300"
              >
                <label
                  className="txt-l"
                  id="SR-CSIRMV1-label"
                >
                  How many dimensions have been breached?
                   
                  <span
                    className="t--req"
                  >
                    Required
                  </span>
                </label>
              </div>
              <div
                className="g"
              >
                <div
                  className="g--12"
                >
                  <div
                    className="ra"
                  >
                    <input
                      checked={false}
                      className="ra-f"
                      id="SR-CSIRMV1-One"
                      name="SR-CSIRMV1"
                      onChange={[Function]}
                      type="radio"
                      value="One"
                    />
                    <label
                      className="ra-l"
                      htmlFor="SR-CSIRMV1-One"
                    >
                      One
                    </label>
                  </div>
                  <div
                    className="ra"
                  >
                    <input
                      checked={false}
                      className="ra-f"
                      id="SR-CSIRMV1-Two"
                      name="SR-CSIRMV1"
                      onChange={[Function]}
                      type="radio"
                      value="Two"
                    />
                    <label
                      className="ra-l"
                      htmlFor="SR-CSIRMV1-Two"
                    >
                      Two
                    </label>
                  </div>
                  <div
                    className="ra"
                  >
                    <input
                      checked={false}
                      className="ra-f"
                      id="SR-CSIRMV1-Three"
                      name="SR-CSIRMV1"
                      onChange={[Function]}
                      type="radio"
                      value="Three"
                    />
                    <label
                      className="ra-l"
                      htmlFor="SR-CSIRMV1-Three"
                    >
                      Three
                    </label>
                  </div>
                  <div
                    className="ra"
                  >
                    <input
                      checked={false}
                      className="ra-f"
                      id="SR-CSIRMV1-More than Three"
                      name="SR-CSIRMV1"
                      onChange={[Function]}
                      type="radio"
                      value="More than Three"
                    />
                    <label
                      className="ra-l"
                      htmlFor="SR-CSIRMV1-More than Three"
                    >
                      More than Three
                    </label>
                  </div>
                  <span />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="g--5 m-v500"
        >
          <div
            className="br br-a200"
          >
            <div
              className="css-19ub58"
              onClick={[Function]}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              style={Object {}}
            >
              <img
                alt=""
                aria-hidden={true}
                height="324"
                src="/static/img/image-upload.png"
                style={
                  Object {
                    "display": "block",
                    "height": "auto",
                    "width": "100%",
                  }
                }
                width="479"
              />
              <input
                accept="image/*"
                multiple={false}
                onChange={[Function]}
                style={
                  Object {
                    "display": "none",
                  }
                }
                type="file"
              />
            </div>
            <div
              className="br br-t200"
              style={
                Object {
                  "position": "relative",
                }
              }
            >
              <button
                className="btn btn--w btn--b"
                onClick={[Function]}
                style={
                  Object {
                    "visibility": "visible",
                    "width": "100%",
                  }
                }
              >
                Upload Photo
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        className="g m-v500 css-1k8tu3j"
      >
        <div
          className="g--9 t--info m-v200 css-hn0yet"
        >
          <span>
            Please fill out 
            <span
              className="t--req"
            >
              required
            </span>
             fields to continue
          </span>
        </div>
        <button
          className="btn g--3"
          disabled={true}
          type="submit"
        >
          Next
        </button>
      </div>
    </form>
  </div>
</div>
`;
